common/src/main/java/io/netty/util/internal/Murmur3.java

```diff
@@ -0,0 +1,464 @@
+            case 2: k1 ^= accessor.getByte(data, offset + 1) << 8;
+            case 1: k1 ^= accessor.getByte(data, offset);
+            h1 ^= murmurHash3Tail(k1);
+            default:
+                break;
+        }
+
+        //----------
+        // finalization
+        h1 ^= len;
+        return murmurHash3Fmix(h1);
+    }
+
+   /**
+    * Designed to yield identical results as
+    * <a href="https://code.google.com/p/smhasher/source/browse/trunk/MurmurHash3.cpp">MurmurHash3_x86_32</a>.
+    * A version was implemented to use longs for unsigned 32-bit operations but performance suffered by about 2x.
+    */
+   @Override
+   public int hashCode(byte[] data, int offset, int len) {
+         int h1 = seed;
+         //----------
+         // body
+         final int numIntsOffset = len & 0xFFFFFFFC;
+         final long adjustedOffset = offset + accessor.byteArrayBaseOffset();
+         final long end = adjustedOffset + numIntsOffset;
+         for (long i = adjustedOffset; i < end; i += 4) {
+             int k1 = accessor.getInt(data, i);
+             k1 *= MurmurHash3_C1;
+             k1 = Integer.rotateLeft(k1, MurmurHash3_ROLT1);
+             k1 *= MurmurHash3_C2;
+
+             h1 ^= k1;
+             h1 = Integer.rotateLeft(h1, MurmurHash3_ROLT2);
+             h1 = (h1 * 5) + MurmurHash3_C3;
+         }
+
+         //----------
+         // tail
+         offset += numIntsOffset;
+         int k1 = 0;
+         switch (len & 3) {
+             case 3: k1 = accessor.getByte(data, offset + 2) << 16;
+             case 2: k1 ^= accessor.getByte(data, offset + 1) << 8;
+             case 1: k1 ^= accessor.getByte(data, offset);
+             h1 ^= murmurHash3Tail(k1);
+             default:
+                 break;
+         }
+
+         //----------
+         // finalization
+         h1 ^= len;
+         return murmurHash3Fmix(h1);
+     }
+
+     private static int murmurHash3Fmix(int h) {
+         h ^= h >>> 16;
+         h *= 0x85ebca6b;
+         h ^= h >>> 13;
+         h *= 0xc2b2ae35;
+         h ^= h >>> 16;
+         return h;
+     }
+
+     private static int murmurHash3Tail(int k1) {
+         k1 *= MurmurHash3_C1;
+         k1 = Integer.rotateLeft(k1, MurmurHash3_ROLT1);
+         return k1 * MurmurHash3_C2;
+     }
+
+    @Override
+    public int hashCode(byte[] bytes) {
+        return hashCode(bytes, 0, bytes.length);
+    }
+
+    @Override
+    public int hashCodeAsBytes(char[] bytes) {
+        return hashCodeAsBytes(bytes, 0, bytes.length);
+    }
+
+    @Override
+    public int hashCodeAsBytes(CharSequence data) {
+        return hashCodeAsBytes(data, 0, data.length());
+    }
+
+    /**
+     * Provides a means of pulling native types out of array type data structures.
+     */
+    interface NativeTypeAccessor {
+        int getInt(byte[] data, long i);
+        int getIntAsBytes(char[] data, long i);
+        int getIntAsBytes(CharSequence data, int i);
```

Comment:
I will add docs if I can get a performant case insensitive case.  The reason to differentiate the name here is because it is not just a regular `getInt` operation.  The MSB in the char is discarded when building the `int`. So for `char` you have to read 4 indexes to build an `int` to be able to have equivalent algorithms between `char[]`, `byte[]`, and case insensitive.